File: Window/RepeatTransformers.m
//
//  RepeatModeTransformer.m
//  Cog
//
//  Created by Vincent Spader on 2/18/08.
//  Copyright 2008 __MyCompanyName__. All rights reserved.
//

#import "RepeatTransformers.h"

#import "Logging.h"

@implementation RepeatModeTransformer {
    RepeatMode repeatMode;
}

+ (Class)transformedValueClass { return [NSNumber class]; }
+ (BOOL)allowsReverseTransformation { return YES; }

- (id)initWithMode:(RepeatMode)r {
    self = [super init];
    if (self) {
        repeatMode = r;
    }

    return self;
}

// Convert from RepeatMode to BOOL
- (id)transformedValue:(id)value {
    DLog(@"Transforming value: %@", value);

    if (value == nil) return nil;

    RepeatMode mode = (RepeatMode) [value integerValue];

    return @(repeatMode == mode);
}

- (id)reverseTransformedValue:(id)value {
    if (value == nil) return nil;

    BOOL enabled = [value boolValue];
    if (enabled) {
        return @(repeatMode);
    } else if (repeatMode == RepeatModeNoRepeat) {
        return @(RepeatModeRepeatAll);
    } else {
        return @(RepeatModeNoRepeat);
    }
}

@end

@implementation RepeatModeImageTransformer

+ (Class)transformedValueClass { return [NSImage class]; }
+ (BOOL)allowsReverseTransformation { return NO; }

// Convert from string to RepeatMode
- (id)transformedValue:(id)value {
    DLog(@"Transforming value: %@", value);

    if (value == nil) return nil;

    RepeatMode mode = (RepeatMode) [value integerValue];

    if (mode == RepeatModeNoRepeat) {
        return [NSImage imageNamed:@"repeatModeOffTemplate"];
    }
    else if (mode == RepeatModeRepeatOne) {
        return [NSImage imageNamed:@"repeatModeOneTemplate"];
    }
    else if (mode == RepeatModeRepeatAlbum) {
        return [NSImage imageNamed:@"repeatModeAlbumTemplate"];
    }
    else if (mode == RepeatModeRepeatAll) {
        return [NSImage imageNamed:@"repeatModeAllTemplate"];
    }

    return nil;
}

@end

File: Window/RepeatTransformers.swift
import Foundation
import AppKit

@objc(RepeatModeTransformer)
class RepeatModeTransformer: ValueTransformer {
    private var repeatMode: RepeatMode
    
    @objc
    init(mode: RepeatMode) {
        self.repeatMode = mode
        super.init()
    }
    
    override class func transformedValueClass() -> AnyClass {
        return NSNumber.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return true
    }
    
    // Convert from RepeatMode to BOOL
    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }
        
        let mode = RepeatMode(rawValue: (value as? NSNumber)?.intValue ?? 0) ?? .noRepeat
        return NSNumber(value: repeatMode == mode)
    }
    
    override func reverseTransformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }
        
        let enabled = (value as? NSNumber)?.boolValue ?? false
        if enabled {
            return NSNumber(value: repeatMode.rawValue)
        } else if repeatMode == .noRepeat {
            return NSNumber(value: RepeatMode.repeatAll.rawValue)
        } else {
            return NSNumber(value: RepeatMode.noRepeat.rawValue)
        }
    }
}

@objc(RepeatModeImageTransformer)
class RepeatModeImageTransformer: ValueTransformer {
    override class func transformedValueClass() -> AnyClass {
        return NSImage.self
    }
    
    override class func allowsReverseTransformation() -> Bool {
        return false
    }
    
    // Convert from string to RepeatMode
    override func transformedValue(_ value: Any?) -> Any? {
        guard let value = value else { return nil }
        
        let mode = RepeatMode(rawValue: (value as? NSNumber)?.intValue ?? 0) ?? .noRepeat
        
        switch mode {
        case .noRepeat:
            return NSImage(named: "repeatModeOffTemplate")
        case .repeatOne:
            return NSImage(named: "repeatModeOneTemplate")
        case .repeatAlbum:
            return NSImage(named: "repeatModeAlbumTemplate")
        case .repeatAll:
            return NSImage(named: "repeatModeAllTemplate")
        default:
            return nil
        }
    }
}
